packages/playwright-core/src/server/dom.ts

```diff
@@ -760,21 +782,27 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
     const { state = 'visible' } = options;
     if (!['attached', 'detached', 'visible', 'hidden'].includes(state))
       throw new Error(`state: expected one of (attached|detached|visible|hidden)`);
-    const info = this._page.parseSelector(selector, options);
-    const task = waitForSelectorTask(info, state, false, this);
     const controller = new ProgressController(metadata, this);
     return controller.run(async progress => {
       progress.log(`waiting for selector "${selector}"${state === 'attached' ? '' : ' to be ' + state}`);
-      const context = await this._context.frame._context(info.world);
-      const injected = await context.injectedScript();
-      const pollHandler = new InjectedScriptPollHandler(progress, await task(injected));
-      const result = await pollHandler.finishHandle();
-      if (!result.asElement()) {
-        result.dispose();
-        return null;
-      }
-      const handle = result.asElement() as ElementHandle<Element>;
-      return handle._adoptTo(await this._context.frame._mainContext());
+      return this._frame.retryWithProgress(progress, selector, options, async (frame, info, continuePolling) => {
+        // If we end up in the same frame => use the scope again, line above was noop.
+        const task = waitForSelectorTask(info, state, false, frame === this._frame ? this : undefined);
+        const context = await frame._context(info.world);
+        const injected = await context.injectedScript();
+        const pollHandler = new InjectedScriptPollHandler(progress, await task(injected));
+        const result = await pollHandler.finishHandle();
+        if (!result.asElement()) {
+          result.dispose();
+          return null;
+        }
+        const handle = result.asElement() as ElementHandle<Element>;
+        try {
+          return await handle._adoptTo(await frame._mainContext());
+        } catch (e) {
+          return continuePolling;
```

Comment:
This is a small behavior change - previously we would fail, but now we'll retry. I guess that's even good, we should add a test :)